File: 1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.py
# TC- O(mn + mnlog(min(m,n)))
# SC- O(mn)

class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        # we need to find sum of any square in O(1) so compute prefix sum
        rows = len(mat)
        cols = len(mat[0])
        prefix_sum = [[0]*(cols+1) for _ in range(rows+1)]
        for row in range(1, rows+1):
            for col in range(1, cols+1):
                prefix_sum[row][col] = mat[row-1][col-1] + prefix_sum[row-1][col]+ prefix_sum[row][col-1] -prefix_sum[row-1][col-1]

        def satisfies(k):
            # checks to see if there is atleast 1 square of side k whose sum<=threshold
            # topleft corner can go from 0 to rows-k+1
            for i in range(rows-k+1):
                for j in range(cols-k+1):
                    total = prefix_sum[i+k][j+k] - prefix_sum[i][j+k]- prefix_sum[i+k][j]+ prefix_sum[i][j]
                    if total<=threshold:
                        return True
            return False

        left, right = 0, min(rows, cols)
        while(left<right):
            mid = left + (right-left+1)//2
            if satisfies(mid):
                left = mid
            else:
                right = mid-1
        return left

File: 1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.cpp
// TC- O(log(min(m, n))* mn* (mid^2))
// SC- O(1)

// class Solution {
// private:
//     bool squareExists(const int& m, const int& n, const int& mid, const vector< vector<int>>& mat, const int& threshold){
//         for(int i=0; i<(m-mid)+1; i++){
//             for(int j=0; j<(n-mid)+1; j++){
//                 int sum = 0;
//                 for(int row=i; row<i+mid; row++){
//                     for(int col=j; col<j+mid; col++){
//                         sum += mat[row][col];
//                     }
//                 }
//                 if(sum<=threshold)
//                     return true;
//             }
//         }
//         return false;
//     }
// public:
//     int maxSideLength(vector<vector<int>>& mat, int threshold) {
//         int m = mat.size(), n = mat[0].size();
//         int left = 0, right = min(m, n);

//         while(left<right){
//             int mid = left+ (right-left+1)/2;

//             if(squareExists(m, n, mid, mat, threshold))
//                 left = mid;
//             else
//                 right = mid-1;
//         }
//         return left;
//     }
// };

// TC- O(log(min(m, n))*mn)
// SC- O(mn)

class Solution {
private:
    bool squareExists(const int& m, const int& n, const int& mid, const vector< vector<int>>& prefix_sum, const int& threshold){
        for(int i=0; i<(m-mid)+1; i++){
            for(int j=0; j<(n-mid)+1; j++){
                // from i to i+mid-1 and j to j+mid-1
                int sum = prefix_sum[i+mid][j+mid] - prefix_sum[i][j+mid]- prefix_sum[i+mid][j] +prefix_sum[i][j];
                if(sum<=threshold)
                    return true;
            }
        }
        return false;
    }
public:
    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        int m = mat.size(), n = mat[0].size();
        int left = 0, right = min(m, n);

        // compute prefix sum
        vector<vector<int>> prefix_sum(m+1, vector<int>(n+1, 0));
        for(int i=1; i<m+1; i++){
            for(int j=1; j<n+1; j++)
                prefix_sum[i][j] = mat[i-1][j-1] + prefix_sum[i-1][j] +prefix_sum[i][j-1] -prefix_sum[i-1][j-1];
        }

        while(left<right){
            int mid = left+ (right-left+1)/2;

            if(squareExists(m, n, mid, prefix_sum, threshold))
                left = mid;
            else
                right = mid-1;
        }
        return left;
    }
};

     
